### PR TITLE
Thread-per-pod model in Kubelet.

### PR DESCRIPTION
**Pull request by wojtek-t**
Change Kubelet so that each pod is being synced in a separate goroutine.

This change is a prerequisite for optimizations of this path, e.g. if the specification for a Pod didn't change since the last update, resyncing of pod should be done only if the probing don't succeed. I'm going to implement those changes in subsequent PR.

Those changes were discussed in #4119

cc @vmarmol @dchen1107

### CODE REVIEW

**Comment by wojtek-t** [comment]
I looked into failures from shippable and it looks really strange:
1. with go 1.3 github.com/GoogleCloudPlatform/kubernetes/pkg/api failed (and I didn't touch this part of code at all)
2. with go 1.4 test-integration.sh failed twice
 In travis all tests are passing, I run test-integration locally 10 times and it passed in all runs. Is it possible that there is something wrong with shippable?

**Comment by vmarmol** [comment]
I'll take a look today. Don't worry about the Shippable failure, we've had issues with it this week and Travis runs the same tests.

**Comment by dchen1107** [comment]
cc/ @yifan-gu

File: pkg/kubelet/pod_workers.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/golang/glog"
+)
+
+type podWorkers struct {
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: nit: can we document the member variables?

File: pkg/kubelet/pod_workers.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+			continue
+		}
+		p.doneTask()
+	}
+}
+
+func (p *podWorkers) UpdatePod(pod api.BoundPod) {
+	uid := pod.UID
+	var podUpdates chan api.BoundPod
+	var exists bool
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: nit: you don't need to pre-declare exists

File: pkg/kubelet/pod_workers.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	if podUpdates, exists = p.podsUpdates[uid]; !exists {
+		// TODO(wojtek-t): Should the channel be unbuffered?
+		podUpdates = make(chan api.BoundPod)
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Make the channel buffered. This will keep one long syncPod() from holding up future updates.

File: pkg/kubelet/pod_workers.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/dockertools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/golang/glog"
+)
+
+type podWorkers struct {
+	podsUpdates map[types.UID]chan api.BoundPod
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: nit: podUpdates (both plural sounds a bit weird :) )

File: pkg/kubelet/pod_workers.go (hunk 1 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Lets also make this class thread-safe.

File: pkg/kubelet/pod_workers.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	}
+}
+
+func (p *podWorkers) managePodLoop(podUpdates <-chan api.BoundPod) {
+	for newPod := range podUpdates {
+		// We may consider passing containers together with the pod from Kubelet.
+		// However, since we currently have cache this shouldn't cause performance
+		// overhead on Docker. Moreover, as long as we run syncPod no matter if
+		// it changes anything, having an old version of "containers" can cause
+		// starting unneded containers.
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: nit: unended

File: pkg/kubelet/pod_workers.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		// TODO(wojtek-t): Revisit it once probing is divided from syncPod.
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: We can probably remove the TODO and the comment, I think you're right and this approach is fine.

File: pkg/kubelet/pod_workers.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		containers, err := p.dockerCache.RunningContainers()
+		if err != nil {
+			glog.Errorf("Error listing containers: %v", err)
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: lets add a bite more info: "while synching pod <uid>"

File: pkg/kubelet/pod_workers.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+			p.doneTask()
+			continue
+		}
+		err = p.syncPodFun(&newPod, containers)
+		if err != nil {
+			glog.Errorf("Error syncing pod, skipping: %v", err)
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: nit: "syncing pod <UID>"

File: pkg/kubelet/pod_workers.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+			record.Eventf(&newPod, "failedSync", "Error syncing pod, skipping: %v", err)
+			p.doneTask()
+			continue
+		}
+		p.doneTask()
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: This gets tricky to handle, consider this (up to you whether to change it):

```Go
for newPod := range channel {
  func() {
    defer p.doneTask()
    if err {
      return
    }
  }()
}
```

File: pkg/kubelet/pod_workers.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Maybe not in this PR, but now that this has a nice API and is separate we should write unit tests for it :)

File: pkg/kubelet/kubelet_test.go [hunk 11]
```diff
@@ -455,15 +457,7 @@ func TestSyncPodsWithTerminationLog(t *testing.T) {
...
-	for {
-		kl.podWorkers.lock.Lock()
-		length := len(kl.podWorkers.workers)
-		kl.podWorkers.lock.Unlock()
-		if length == 0 {
-			return
-		}
-		time.Sleep(time.Millisecond * 100)
-	}
+	kl.podWorkers.waitGroup.Wait()
```
Comment (vmarmol) [inline pkg/kubelet/kubelet_test.go]: WDYT if instead of this as part of newTestKubelet we create a podWorkers where the syncPodFun is:

```Go
func() {
  syncPod()
  waitgroup.Done()
}
```

Just trying to find a way to keep the test-only code outside of the production code.

File: pkg/kubelet/pod_workers.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Both done.

File: pkg/kubelet/pod_workers.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: hmm - I don't know how to do this. I want to have podUpdates available after the if. Is it possible to not declare exists then?

File: pkg/kubelet/pod_workers.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Done.

File: pkg/kubelet/pod_workers.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Done.

File: pkg/kubelet/pod_workers.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Done.

File: pkg/kubelet/pod_workers.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Thanks! Done.

File: pkg/kubelet/pod_workers.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Done.

File: pkg/kubelet/pod_workers.go (hunk 8 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Sorry - I don't understand this comment.

File: pkg/kubelet/pod_workers.go (hunk 9 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Nice - this is something I wanted to do but I didn't know how to do it in go.

File: pkg/kubelet/pod_workers.go (hunk 10 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Added a TODO - will do it after this PR.

File: pkg/kubelet/kubelet_test.go (hunk 11 above)
Comment (wojtek-t) [inline pkg/kubelet/kubelet_test.go]: I slightly changed your suggestion, but I agree that it's much cleaner when there is no test code in podWorkers class. Hopefully you will also like it more.

File: pkg/kubelet/pod_workers.go (hunk 10 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Addressed in #4839

File: pkg/kubelet/pod_workers.go (hunk 8 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Sorry I left out the last part :) "syncing pod <UID>", lets add the pod's UID to the error

File: pkg/kubelet/pod_workers.go (hunk 10 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Thanks! Will take a look at that too.

File: pkg/kubelet/pod_workers.go (hunk 2 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Don't you only use exists in the if statement? I agree that podUpdates needs to be pre-declared.

**Comment by vmarmol** [comment]
Looks good, only those two nits. We'll have this merged today :)

File: pkg/kubelet/pod_workers.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Sorry - I might be doing it wrong, but that doesn't work for me. If I don't declare "exists" and change "=" to ":=" in the if it doesn't work for me (i.e. it hangs forever in tests so it seems that it doesn't use the predeclared podUpdates var).

File: pkg/kubelet/pod_workers.go (hunk 8 above)
Comment (wojtek-t) [inline pkg/kubelet/pod_workers.go]: Done.

**Comment by wojtek-t** [comment]
PTAL

File: pkg/kubelet/pod_workers.go (hunk 2 above)
Comment (vmarmol) [inline pkg/kubelet/pod_workers.go]: Aaaaah yes you're correct. I just realized that it will overwrite both with `:=` and thus podUpdates would be empty.

**Comment by vmarmol** [comment]
LGTM, will merge when green. Thanks @wojtek-t!

**Comment by ixdy** [comment]
FYI, this was reverted by #4867 because it broke e2e tests on gce.

File: pkg/kubelet/kubelet_test.go [hunk 12]
```diff
@@ -293,7 +301,7 @@ func TestKillContainerWithError(t *testing.T) {
 		Err:           fmt.Errorf("sample error"),
 		ContainerList: append([]docker.APIContainers{}, containers...),
```
Comment (satnam6502) [inline pkg/kubelet/kubelet_test.go]: I know this line is not part of this PR, but why write `append([]docker.APIContainers{}, containers...)` when you could write `containers` ?
